### ISSUE TITLE
Concurrent Update Exception in ChoETL.OrderedDictionary

### ISSUE THREAD

**Issue by gjvanvuuren** [Cinchoo/ChoETL#200]
I'm processing large Avro files using `ChoETL.JSON.NETStandard` version 1.2.1.47. The Avro file contains different JSON structures, representing different messages. Using multitasking, each record (or message) is processed separately using the `ChoJSONReader`. Each message processing Task basically looks like this:

```

        private async Task<int> ExtractMessageXYZPartAsync(MessageHeader msgHeader, MessageBody msgBody, string msgSrc)
        {
            var jsonBody = msgBody.JsonBody;

            var list = this.messages.GetOrAdd
            (
                nameof(MessageXYZ),
                new ConcurrentBag<MessageXYZ>()
            ) as ConcurrentBag<MessageXYZ>;

            using
            (
                var r = new ChoJSONReader(new JsonTextReader(new StringReader(jsonBody)))
                    .WithField(...)
                    .WithField(...)
                    ...

                    .Configure
                    (
                        cfg => new ChoJSONRecordConfiguration
                        {
                            Culture = CultureInfo.InvariantCulture,
                            FlattenNode = true,
                            ErrorMode = ChoErrorMode.IgnoreAndContinue,
                            JsonSerializerSettings = new JsonSerializerSettings
                            {
                                DateParseHandling = DateParseHandling.None
                            }
                        }
                    )
            )
            {
                try
                {
                    await foreach(var msgPart in r.ToAsyncEnumerable())
                    {
                        var msg = new MessageXYZ()
                        {
                            ...
                        };

                        list.Add(msg);
                    }
                }
                catch(Exception ex)
                {
                    ex.Data.Add("Body", jsonBody);
                    ExceptionDispatchInfo.Capture(ex).Throw();
                }
            }

            return list.Count();
        }
```

Unfortunately, I run into concurrency issues with ChoETL. Here's a partial example of a typical stack trace:

```
Message: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.

System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.ObjectModel.KeyedCollection`2.AddKey(TKey key, TItem item)
   at System.Collections.ObjectModel.KeyedCollection`2.InsertItem(Int32 index, TItem item)
   at System.Collections.ObjectModel.Collection`1.Add(T item)
   at ChoETL.OrderedDictionary`2.Add(TKey key, TValue value)
   at ChoETL.ChoDynamicObject.SetPropertyValue(String name, Object value)
   at CallSite.Target(Closure , CallSite , Object , ChoJSONRecordConfiguration )
   at ChoETL.ChoJSONRecordConfiguration.Validate(Object state)
   at ChoETL.ChoJSONRecordReader.AsEnumerable(IEnumerable`1 jObjects, TraceSwitch traceSwitch, Func`2 filterFunc)+MoveNext()
   at ChoETL.ChoJSONRecordReader.AsEnumerable(Object source, Func`2 filterFunc)+MoveNext()
   at ChoETL.ChoEnumeratorWrapper.BuildEnumerable[T](Func`1 moveNext, Func`1 current, Action dispose)+MoveNext()
   at System.Linq.AsyncEnumerable.AsyncEnumerableAdapter`1.MoveNextCore() in /_/Ix.NET/Source/System.Linq.Async/System/Linq/Operators/ToAsyncEnumerable.cs:line 68
   at System.Linq.AsyncIteratorBase`1.MoveNextAsync() in /_/Ix.NET/Source/System.Linq.Async/System/Linq/AsyncIterator.cs:line 70
   at System.Linq.AsyncIteratorBase`1.MoveNextAsync() in /_/Ix.NET/Source/System.Linq.Async/System/Linq/AsyncIterator.cs:line 75
   at CompanyX.DomainY.MessageCaptureApp.AvroBlobObject.ExtractMessageXYZPartAsync(MessageHeader msgHeader, MessageBody msgBody, String msgSrc) in 
```

I don't understand why there should be concurrency issues here. Do multiple instances of ChoJSONReader share a common configuration space?

**Comment by Cinchoo** [comment]
I'm not sure about it. let me look at it.

In the mean time, pls set `ChoDynamicObjectSettings.DictionaryType = DictionaryType.Regular` at the main() and see if this works.

**Comment by gjvanvuuren** [comment]
Thanks for the feedback. I've set `ChoDynamicObjectSettings.DictionaryType = DictionaryType.Regular` in the constructor of the message processing class, but there's no difference. I still get the concurrency exception.

**Comment by Cinchoo** [comment]
Found the issue, applied fix. Released v1.2.1.48-beta2. Give it a try and let me know.

**Comment by gjvanvuuren** [comment]
Perfect! Problem solved, thanks!
